### PR TITLE
Fix an issue where closure rewriting required class internals 

### LINKED ISSUE #1174: Returning a struct from a closure results in "__wrap is not a function" on that struct
I'm on wasm-bindgen@0.2.31 and getting "__wrap is not a function" when calling a closure which returns a struct that is not used anywhere else. 

Here's the smallest example of the bug, I could make: 

```rust
type ClosureType = FnMut() -> BadStruct;

#[wasm_bindgen]
pub struct ClosureHandle(Closure<ClosureType>);

#[wasm_bindgen]
pub struct BadStruct {}

#[wasm_bindgen]
pub fn closure_returner() -> Result<Object, JsValue> {
    let o = Object::new();

    let some_fn = Closure::wrap(Box::new(move || BadStruct {}) as Box<ClosureType>);
    Reflect::set(
        &o,
        &JsValue::from("someKey"),
        &some_fn.as_ref().unchecked_ref(),
    )
    .map_err(|_| "Unable to set property")?;
    Reflect::set(
        &o,
        &JsValue::from("handle"),
        &JsValue::from(ClosureHandle(some_fn)),
    )
    .map_err(|_| "Unable to set closure handle")?;

    Ok(o)
}
```

Typescript:
```typescript
import("../../lib-web/pkg").then(module => {
    const c = module.closure_returner();
    c.someKey(); // Error occurs here
});
```

Currently this works as a workaround:
```
#[wasm_bindgen]
pub fn _dont_call() -> BadStruct {
    unimplemented!();
}
```

Edit: Related to #949, but not exactly the same.

### PR DESCRIPTION
**Pull request by alexcrichton**
Surfaced through previous sanity-checking commits, this reorders some
internal operations to...

Closes #1174

### CODE REVIEW

**Comment by alexcrichton** [comment]
@fitzgen I'm curious if you have thoughts on this, we've had a good number of bugs about the ordering of operations in `finalize` in wasm-bindgen, I feel like we're constantly playing some sort of weird code golf to figure out the right ordering.

Do you have thoughts on how we might better structure this and/or fix this for good in terms of JS generation?

File: crates/cli-support/src/js/mod.rs [hunk 1]
```diff
@@ -478,6 +484,10 @@ impl<'a> Context<'a> {
         self.rewrite_imports(module_name);
         self.update_producers_section();
 
+        // Cause any future calls to `should_write_global` to panic, making sure
+        // we don't ask for items which we can no longer emit.
+        self.globals_written = true;
```
Comment (fitzgen) [inline crates/cli-support/src/js/mod.rs]: Alternatively, `exposed_globals` could be an `Option<...>` and then we can have `should_write_global` do `self.exposed_globals.as_mut().unwrap().insert(name)` and then here do `self.exposed_globals.take().unwrap()`.

This is equivalent to having a bool flag (which becomes the option discriminant in this scheme) but it is *impossible* to forget to check the flag in `should_write_global` or anything else that wants to touch `exposed_globals` thanks to the type system.

File: crates/cli-support/src/js/mod.rs [hunk 2]
```diff
@@ -56,7 +56,7 @@ pub struct Context<'a> {
     /// wasm-bindgen emits.
     pub direct_imports: HashMap<&'a str, (&'a str, &'a str)>,
 
-    pub exported_classes: HashMap<String, ExportedClass>,
+    pub exported_classes: Option<HashMap<String, ExportedClass>>,
```
Comment (fitzgen) [inline crates/cli-support/src/js/mod.rs]: Ah ok, this second commit is exactly the pattern that I was suggesting for the first one :-p

File: crates/cli-support/src/js/mod.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/cli-support/src/js/mod.rs]: 👍

**Pull request merged**